### ISSUE TITLE
Candidates::softmax(None) does a gratuitous full-vocab sort (per-token perf regression)

### ISSUE THREAD

**Issue by claudeopusagora** [mdegans/drama_llama#33]
## Symptom
`whodunit_verdict` (structured output, `locally_typical` default) runs ~11 min in the debug test suite and ~30 tok/s in release on Qwen3.6-35B-A3B — far slower than a 3B-active MoE on a 3090 should be (~60–100 tok/s target). Flat per-token cost, independent of context length.

## Profiling (samply, opt-3 build, addr2line-symbolicated)
~**48% of every token** is in `partial_sort-0.2.0/src/lib.rs`, sorting the full **248,320**-token vocab. GPU forward is only ~22% (`libcuda`); the sampler sort is the wall.

## Root cause
`Candidates::softmax(&self, k: Option<_>)` (`src/candidates.rs:583`):
```rust
let k = k.map(|k| k.min(self.len())).unwrap_or(self.len()); // None -> full vocab
let mut new = self.sort(Sorted::ByLogit { k }).truncate(k);  // O(n log n) heapsort of 248k
```
`locally_typical` (the default sampling mode) calls `self.softmax(None)` **every token**, so we pay a full-vocab heapsort per token. But **softmax is order-independent** — it only needs the max for numerical stability, which is an **O(n) scan**, not an O(n·log n) sort.

Likely a **refactoring regression** — see the `// This has just been refactored so it should be checked against the original` comment near the top_p block (`src/candidates.rs:712`). Exposed now by this model's unusually large 248k vocab; smaller-vocab models (128k–151k) hid it. `git blame` on `softmax` should date it.

## Fix
Make `softmax` obtain the max via a linear scan instead of a full sort. Before touching it, verify no caller relies on `softmax` returning sorted candidates (`choose_candidate`, `locally_typical`, top_p all call it).

## Impact
Removes ~48% of per-token cost on the default sampling path; ~6.6× on debug (the whole `just test full` suite is debug-bound by this).

_Profiled on CUDA (RTX 3090). Mac cross-check pending (see #2)._

**Comment by mdegans** [comment]
## Mac/Metal cross-check (the one this issue was waiting on)

Confirmed on Apple Silicon, 2026-07-14. Same tax, and Mike independently localized the time to `sort` before seeing this issue.

**But the arithmetic doesn't close, and I think there is a *second* cost here.**

Two runs on the same machine, same model family, same sampling policy:

| run | grammar? | tok/s |
|---|---|---|
| `bench.py --model a3b --backend llama-cpp` (blallama, Qwen3.6-35B-A3B-UD-Q4_K_S) | none | **24 tok/s** |
| `just example whodunit` (Qwen3.6-35B-A3B-UD-IQ4_XS, `structured_output`) | JSON | **~15 min wall** |

The sampling policy is the same in both: I diffed the sidecars, and `models/model.sampling.toml` and `~/models/gguf/Qwen3.6-35B-A3B-UD-Q4_K_S.sampling.toml` are the same policy — `LocallyTypical p=0.5` + identical repetition config. So **bench pays this issue's full-vocab sort too**, once per token, and still turns in 24 tok/s.

That bounds what this issue can explain. At ~48% of per-token cost, fixing `softmax(None)` is worth roughly **2×** — real and worth doing, but it cannot account for whodunit being slower by more than an order of magnitude at what is (per Mike) *not* runaway generation.

**So the grammar path appears to carry a second, separate cost on top of this one.** The only thing whodunit adds over bench is the JSON grammar. Prime suspect: the lazy grammar check (#28) samples first and checks the sampled token, and **on rejection re-runs with a full O(vocab) mask** — which re-runs the chain and therefore triggers *another* `softmax(None)` full-vocab sort. JSON is a tight constraint, so N rejections per token ⇒ N+1 sorts per token, and the fix in this issue would then be multiplied, not merely additive.

That's a hypothesis, not a measurement — I have not profiled the Mac, and I don't know whodunit's actual output token count (if it emitted ~20k tokens, 15 min at 24 tok/s is simply arithmetic and there's no second cost at all). Both are cheap to settle:

1. Count whodunit's generated tokens. Rules the arithmetic in or out immediately.
2. `grammar_stats_enabled()` instrumentation already exists in `sample.rs` — count fallback re-runs per token. If it's >0 on average, the second cost is confirmed and `softmax` is being called several times per token.

Do (1) first; it's free and it might make (2) unnecessary.

Filed from the v0.8.0 pre-publish validation pass. Not blocking publish — it is a perf issue, not a correctness one.

**Comment by mdegans** [comment]
## Correction and a sharper suspect — the lazy check's fallback, not the sort

Mike pointed out that `Candidates` caches its sort state (`sort()` early-returns when `cached_k >= k`), so "we re-sort N times per token" should be impossible. He's right, and chasing that is what found the actual shape.

### The sort cache is being routed around, not failing

`sample_token` (`src/sample.rs:1242`):
```rust
let snapshot = if lazy || !banned.is_empty() {
    Some((*rng, *mu, candidates.clone()))     // <-- cloned BEFORE apply_modes
};
let filtered = apply_modes(candidates, opts, rng, mu, model, lazy);
...
if !legal {
    let filtered = apply_modes(saved.clone(), opts, rng, mu, model, false);   // cold
    chosen = choose_candidate(rng, filtered.softmax(None))...
}
```
The snapshot is taken **pre-fold**, so `saved.sort_state == Sorted::No`. The rejection path sorts a *fresh copy*, and the cache cannot fire — it isn't the same set. Working as designed, defeated by the clone.

### Two costs, and only one of them is this issue

1. **~3 MB memcpy per token, unconditionally.** `TokenData` is 12 bytes; 248,320 × 12 ≈ 3 MB per clone. The guard is `lazy || !banned.is_empty()`, and `emit_specials_ban` defaults to **true** (`session/mod.rs:1369`) — so *every* `Session` token clones the full candidate vector, grammar or not. bench pays it and still does 24 tok/s, so it's a tax, not the wall.
2. **A second, cold, full-vocab sort per rejected token.** Real, and it multiplies this issue's fix rather than adding to it.

### But the sorts still don't add up, and I want to be clear about that

One rejection ⇒ two sorts, not N. At ~48% of a token, doubling the sort is ~1.5× overall. That does not reach the >10× gap between bench (24 tok/s, no grammar) and whodunit (~15 min).

**The bigger suspect is the fallback's O(vocab) grammar sweep.** `apply_modes(..., lazy=false)` runs the full `par_iter` over all 248,320 candidates calling `accepts_bytes` on each — the exact sweep #28 was built to eliminate. The lazy check only avoids it **when the sampled token is accepted**. A JSON grammar is a tight constraint, so if the rejection rate is high, #28's premise doesn't hold for structured output and we pay the full sweep on most tokens *anyway*, plus a second 3 MB clone and a cold sort.

That would make #28 a win for loose grammars and a near-no-op (with overhead) for tight ones — which is a hypothesis about the *design*, not a bug in the implementation.

### The measurement is already in the tree

`grammar::record_lazy(legal, t0.elapsed())` records accept/reject per token behind `grammar_stats_enabled()`. **Get the rejection rate before optimizing anything.**

- Low rejection rate ⇒ #28 is working, the fallback is rare, and this issue's `softmax` fix is the whole story.
- High rejection rate ⇒ the fallback is the steady state, and the fix is upstream of `softmax`: keep the sorted set (snapshot post-sort, or make the rerun reuse the sorted candidates so the cache fires), and reconsider whether a rejected sample should trigger a *full* vocab sweep or an incremental repair.

Also still worth ruling out first, for free: **count whodunit's output tokens.** If it emitted ~20k, 15 min at 24 tok/s is arithmetic and there is no mystery.

**Comment by mdegans** [comment]
## Correction — I had the mechanism wrong, and the right one makes the fix cleaner

Mike caught it: `Candidates` derives `Clone`, so `sort_state` **is** carried through `candidates.clone()`. My previous comment claimed the clone "discards the sort state." It does not. Retracting that.

The real chain, per token:

1. `Candidates::from_logits` → `from_iter` → **`from_vec`**, which does an O(n) scan and finds ids ascending ⇒ `sort_state = Sorted::ById { k: n }`.
2. `candidates.clone()` for the snapshot — faithfully preserves `ById { n }`.
3. Pass 1: `softmax(None)` → `sort(Sorted::ByLogit { k: n })`. Cached axis is **ById**, query is **ByLogit** ⇒ **cache miss** ⇒ full 248k sort.
4. Fallback: `apply_modes(saved.clone(), …)` — `saved` is still `ById { n }` ⇒ `softmax(None)` ⇒ **miss again** ⇒ a second full 248k sort.

So the clone preserves the sort state perfectly. It preserves the **wrong axis**. `ById` is worthless to a `ByLogit` query, and the snapshot is taken *before* the only sort that would have helped — there is nothing useful to clone yet.

### That makes the clone the fix rather than the problem

**Sort `ByLogit` once, up front, before taking the snapshot.** Then `Clone` carries `ByLogit { n }` into `saved`, and *both* `softmax(None)` calls are cache hits. One sort per token instead of two, and the existing sort cache does exactly what it was built to do.

Should be behavior-neutral: pass 1 already ends `ByLogit` before `choose_candidate`, and so does the fallback, so the final ordering — and therefore the RNG-draw → token mapping — is unchanged. Worth confirming against `tests/regression.rs` (greedy, so insensitive) *plus* one sampled run at a fixed seed (which is the case that would actually notice).

### Separate free win in the same path

`from_logits` builds ids as `(0..).zip(logits)` — **ascending by construction**. So `from_vec`'s O(n) sort-order scan runs every token to rediscover a fact the constructor already knows, and always concludes `ById`. `Candidates::from_vec_unchecked_full(data, Sorted::ById { k: n })` skips it. That is the *checked-invariant* constructor, not the unsound one — no soundness surface, just a linear scan of 248k elements deleted from every token.

### Still the thing to measure first

None of this changes the priority: **get the lazy-check rejection rate** (`grammar_stats_enabled()` → `record_lazy`) before optimizing. If rejections are rare, the fallback almost never runs and its second sort is irrelevant; if they're common, #28's premise doesn't hold for tight JSON grammars and the O(vocab) `accepts_bytes` sweep — not the sort — is the wall.

And still free to rule out first: count whodunit's actual output tokens.

**Comment by mdegans** [comment]
Leads from the 2026-07-16 config/state-split session (code reading, not yet profiled):

- **The 0.8.0 default delta is repetition, not the mode**: `SamplerConfig::default()` turned the repetition penalty on by default in 0.8.0. The pass's entry `sort(ById)` is a no-op (`from_logits` caches id order), so it shouldn't be the *sorting* hotspot, but it's per-token work whodunit didn't pay pre-0.8. A/B: `repetition: None`.
- **Bare `[LocallyTypical]` = two full-vocab sorts per token**: `softmax(None)` re-sorts ~150k candidates id→logit via `partial_sort(k = len)` (heap selection — worse constants than `sort_unstable` at k ≈ n), then the typicality sort. Cheap experiments: (a) branch to `sort_unstable_by` when `k == len`; (b) default chain `TopK → LocallyTypical` so downstream sorts are k-sized; (c) read sampling defaults from GGUF metadata when present (Qwen ships them) — if 15 min collapses to ~30 s on any of these, that's the answer.
- Re-time on the v0.8.0 branch first: the split landed there homes the DFA cache in config (warm across calls, was cold per call) — if the second suspect was grammar-side, it may already be better.

**Comment by mdegans** [comment]
Measured (2026-07-16, M2 Max, Qwen3.6-35B-A3B): whodunit_verdict ~900s baseline → **395s** after flipping the lazy_grammar default on (f35444c, #28 step 3 — Session had run the masked full-vocab sweep since #28 landed) → **92s** under `--cargo-profile release` (the test suite runs opt-level 0; nextest doesn't override the cargo test profile). ~10× stacked; full suite 2027s → 1777s (93/93 green).

Remaining levers from the earlier comment (top-k before LocallyTypical, GGUF-metadata sampling defaults, sort_unstable at k==len) are now polish rather than rescue — the debug-profile sampler was the bulk of the residual. Candidate standing fix for suite wall-clock: `[profile.test] opt-level = 1` in Cargo.toml or a release-profile `just test full` variant.

**Comment by mdegans** [comment]
## Closing: acute symptom resolved by two landed changes

Revisiting on the v0.8.0 branch, the full-vocab softmax sort this issue
profiled no longer runs on the default path, and the >10× whodunit wall
it was entangled with was a different cost entirely. Two changes closed it:

1. **The real wall was the lazy-grammar O(vocab) `accepts_bytes` sweep,
   not the sort.** Flipping `lazy_grammar` on (`f35444c`, #28 step 3) took
   `whodunit_verdict` 900s → 395s (→ 92s under `--cargo-profile release`).
   The softmax sort was never more than ~2× of a *token*; it could not
   account for the order-of-magnitude gap.

2. **`TopK{1024}` pre-cut ahead of `locally_typical` by default**
   (`16dde17`, 2026-07-18). With top-k sorting+truncating to 1024 first,
   `locally_typical`'s `softmax(None)` now hits the cached
   `ByLogit{1024}` no-op branch and computes probabilities over 1024
   entries, not the full 248k vocab. The full-vocab sort is gone from the
   default chain.

## What remains is polish, not rescue — and deliberately deferred

- A **bare, non-default chain** (`[LocallyTypical]` from a sidecar, or a
  caller building their own chain with no `TopK` pre-cut) still eats one
  full-vocab softmax sort per token. Latent inefficiency, not a live
  regression on the default path.
- Micro: `partial_sort(k == len)` is heap-select where `sort_unstable_by`
  has better constants; and `from_vec`'s per-token O(n) id-order rescan is
  skippable (`from_logits` builds ids ascending). Both tiny.

The tempting "real" fix — splitting `softmax`'s contract so it computes
probabilities in O(n) without sorting — is **not worth it**: the
distribution sampler (`choose_candidate`) walks candidates *in order*, so
its sort is a **seed-reproducibility anchor**, not a correctness
requirement. Removing it wholesale would silently shift every seeded
(non-greedy) output and force a re-bless of the `#[ignore]` model tests —
a large blast radius to reclaim a sort that is already bounded to ≤1024 on
the default path. Not a good trade.

Closing as addressed. If a bare-chain / large-vocab consumer ever makes
the residual sort a measured hotspot again, the O(n)-softmax split is the
lever — with the reproducibility decision made explicitly at that point.

🤖 Closed by Claude (Opus 4.8) via Claude Code, with Mike at the wheel.
